metaflow/plugins/argo/argo_events.py

```diff
@@ -0,0 +1,95 @@
+import json
+import os
+import sys
+import time
+import urllib
+import uuid
+from datetime import datetime
+
+from metaflow.exception import MetaflowException
+
+from metaflow.metaflow_config import ARGO_EVENTS_WEBHOOK_URL
+
+
+class ArgoEventException(MetaflowException):
+    headline = "Argo Event Exception"
+
+
+class ArgoEvent(object):
+    def __init__(
+        self, name, url=ARGO_EVENTS_WEBHOOK_URL, payload={}, access_token=None
+    ):
+        # TODO: Introduce support for NATS
+        self._name = name
+        self._url = url
+        self._payload = payload
+        self._access_token = access_token
+
+    def add_to_payload(self, key, value):
+        self._payload[key] = str(value)
+        return self
+
+    def publish(self, payload={}, force=False, ignore_errors=False):
+        # Publish event iff forced or running on Argo Workflows
+        if force or os.environ["ARGO_WORKFLOW_TEMPLATE"]:
+            try:
+                headers = {}
+                if self._access_token:
+                    # TODO: Test with bearer tokens
+                    headers = {"Authorization": "Bearer {}".format(self._access_token)}
+                # TODO: do we need to worry about certs?
+
+                # Use urllib to avoid introducing any dependency in Metaflow
+                request = urllib.request.Request(
+                    self._url,
+                    method="POST",
+                    headers={"Content-Type": "application/json", **headers},
+                    data=json.dumps(
+                        {
+                            "name": self._name,
+                            "payload": {
+                                # Add default fields here...
+                                "name": self._name,
+                                "id": str(uuid.uuid4()),
+                                "timestamp": int(time.time()),
+                                "utc_date": datetime.utcnow().strftime("%Y%m%d"),
+                                "generated-by-metaflow": True,
+                                **self._payload,
+                                **payload,
+                            },
+                        }
+                    ).encode("utf-8"),
+                )
+                retries = 3
+                backoff_factor = 2
+
+                for i in range(retries):
+                    try:
+                        urllib.request.urlopen(request, timeout=10.0)
+                        print(
+                            "Argo Event (%s) published." % self._name, file=sys.stderr
```

Comment:
Publish notices (success or failure): should we be printing these by default.  They were very useful in my testing / infra setup work.  But doing this routinely for happy path is not ideal.

Reply:
Yeah - we can turn them off and only print the failures later. these go to stderr so hopefully less cloggy.